projects/laji/src/app/+project-form/form/document-form/document-form.component.ts

```diff
@@ -1,66 +1,311 @@
+  readonly: Readonly;
+  isAdmin: boolean;
+  namedPlaceHeader: string[];
+  locked?: boolean;
+  editingOldWarning?: string;
+}
+
+function isSaneViewModel(any: ViewModel): any is SaneViewModel {
+  return typeof any !== 'string' && any && !!any?.form;
+}
+
+type ViewModel = SaneViewModel | FormError;
+
+interface DocumentAndHasChanges {
+  document: Document;
+  hasChanges: boolean;
+}
 
 @Component({
-  template: `
-    <laji-document-form
-      [formId]="form.id"
-      [documentId]="documentID"
-      [showHeader]="!form.options?.mobile"
-      [showShortcutButton]="!form.options?.mobile"
-      (missingNamedplace)="goBack()"
-      (success)="onSuccess()"
-      (error)="goBack()"
-      (leave)="goBack()"
-      (goBack)="goBack()"
-    >
-    </laji-document-form>
-  `,
-  selector: 'laji-project-form-document-form',
+  selector: 'laji-document-form',
+  templateUrl: './document-form.component.html',
+  styleUrls: ['./document-form.component.css'],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class DocumentFormComponent {
-  @Input() form: Form.SchemaForm;
+export class DocumentFormComponent implements OnInit, OnDestroy {
+  @ViewChild(LajiFormComponent) lajiForm: LajiFormComponent;
+  @ViewChild('saveAsTemplate', { static: true }) public templateModal: ModalDirective;
+
+  @Input() formID: string;
   @Input() documentID: string;
+  @Input() namedPlaceID: string;
+  @Input() template: boolean;
 
-  npLoaded$ = new BehaviorSubject<boolean>(false);
-  np: NamedPlace;
+  @LocalStorage('tmpDocId', 1) private tmpDocId: number;
 
-  @ViewChild(_DocumentFormComponent) documentComponent: _DocumentFormComponent;
+  vm$: Observable<ViewModel>;
+  private locked$: Observable<boolean | undefined>;
+  private formData$: Observable<any>;
+  private formDataChange$ = new ReplaySubject<any>();
+  private hasChanges$: Observable<boolean>;
+  private onSaved$ = new ReplaySubject<void>();
 
-  @Input() set namedPlace(namedPlace: NamedPlace) {
-    this.lajiFormDocumentFacade.useNamedPlace(namedPlace, this.form.id);
-    this.npLoaded$.next(true);
-    this.np = namedPlace;
-  }
+  private vm: SaneViewModel;
+  private isFromCancel = false;
+  private confirmLeave = true;
+  private saving = false;
+  private publicityRestrictions: Document.PublicityRestrictionsEnum;
+
+  validationErrors: any;
+  templateForm: TemplateForm = {
+    name: '',
+    description: '',
+    type: 'gathering'
+  };
+  documentForTemplate: any = {};
+  touchedCounter$: Observable<number>;
+
+  isFormError = isFormError;
+  isSaneViewModel = isSaneViewModel;
+  errors = FormError;
+
+  private vmSub: Subscription;
+  private saveTmpSub: Subscription;
 
   constructor(
     private router: Router,
     private localizeRouterService: LocalizeRouterService,
     private browserService: BrowserService,
     private route: ActivatedRoute,
-    private namedPlacesService: NamedPlacesService,
-    private lajiFormDocumentFacade: LajiFormDocumentFacade,
```

Comment:
I moved the business logic to a facade. Maybe it's a bit more hygienic now, since the component now only handles communicating between the business logic (facade) and the view. Though I'm not entirely sure if the navigation should be considered business logic or view logic... Also, the facade returns a view model, so the idea of it handling purely business is somewhat artificial. But I feel that in component-oriented modern web UI techniques trying to isolate the business and view isn't necessarily very productive, as the UI and the business is often inherently closely coupled and too much abstraction can be counter productive.